### ISSUE TITLE
[performance] fake_full_tx in TransactionBuilder should use static data instead of cryptography

### ISSUE THREAD

**Issue by CantTouchDis** [Emurgo/cardano-serialization-lib#213]
I have been playing around building transaction with the TransactionBuilder (version `8.1.0` and now `9.1.0`) and stumbled across a problem with `add_change_if_needed`. For some reason this function takes a very long time in my browser. (brave)

I am not a webdev and just wanted to try out if I could get it to work, so it might very well be a problem with my code.

```js
  const txBuilder = S.TransactionBuilder.new(
    // all of these are taken from the mainnet genesis settings
    // linear fee parameters (a*size + b)
    S.LinearFee.new(S.BigNum.from_str('44'), S.BigNum.from_str('155381')),
    // minimum utxo value
    S.BigNum.from_str('1000000'),
    // pool deposit
    S.BigNum.from_str('500000000'),
    // key deposit
    S.BigNum.from_str('2000000'),
    // max_value_size
    300,
    // max_tx_size
    8000
  );
  // add the input
  txBuilder.add_input(
    bestUtxo.output().address(),
    bestUtxo.input(),
    bestUtxo.output().amount()
  );
  // add the input
  txBuilder.add_output(...);
  // calculate change (THIS TAKES FOR AGES)
  txBuilder.add_change_if_needed(
    changeAddr
  );
  var body = txBuilder.build();

  var transaction = S.Transaction.new(
    body,
    S.TransactionWitnessSet.new(),
    // auxData
  );
  ... sign, set witness and submit
```

Info: I did all of this with testnet addresses, have not tried it yet with mainnet ones. `bestUtxo` in the above is just one `TransactionUnspentOutput` with enough ada, returned by nami wallet

Ignoring the parameters `max_value_size` and `max_tx_size` as it happened in the previous version for me as well, what could be the issue?

Callgraph profiling in my browser:

![ChangeIfNeeded](https://user-images.githubusercontent.com/5838555/134809806-81f1d7b3-17b2-436a-b87d-12140dfabd40.png)

If you need any more information please let me know.

**Comment by CantTouchDis** [comment]
With the 9.1.0 lib I got readable names in the Profiler view:
![Profile9 1 0](https://user-images.githubusercontent.com/5838555/134814798-9d3d786e-b260-4a21-90f7-6ff4031fd7db.png)

**Comment by CantTouchDis** [comment]
I just tried it in my firefox and got the same result for this minimal example code: (assuming S is the loaded module)
```js
function hexToBytes(hex) {
  var bytes = new Uint8Array(hex.length / 2);
    for (var c = 0; c < hex.length; c += 2)
      bytes[c / 2] = parseInt(hex.substr(c, 2), 16);
    return bytes;
}

const changeAddr = S.Address.from_bytes(hexToBytes('00fbb9d8bfc1acef684e39e5b4f7db587a0a8078f1a773321a8a0c4439a7e68f2739efdbe66b3a61250b95048415c6d78bdb0ec2fad231e594'));
const receiver = S.Address.from_bech32('addr_test1qqvhrap0cye6c5fmhlknuv8jzllyy33hwvc86q6qvamzungxu2hyfhlkwuxupa9d5085eunq2qywy7hvmvej456flknsyrt82q');
const utxos = ["82825820b7e8599a28d28122bfb2b9d793883c56f9f5d697ccc08f3de545bcd40c02ce1f0182583900fbb9d8bfc1acef684e39e5b4f7db587a0a8078f1a773321a8a0c4439a7e68f2739efdbe66b3a61250b95048415c6d78bdb0ec2fad231e5941a39cc8510", "8282582023d14445d0a7fe9a1a66c9fefa084f8b875ca0aa600e86ed28d50d0618021c170182583900fbb9d8bfc1acef684e39e5b4f7db587a0a8078f1a773321a8a0c4439a7e68f2739efdbe66b3a61250b95048415c6d78bdb0ec2fad231e5941a00102bae"
].map(u => S.TransactionUnspentOutput.from_bytes(hexToBytes(u)));
const txBuilder = S.TransactionBuilder.new(
    // all of these are taken from the mainnet genesis settings
    // linear fee parameters (a*size + b)
    S.LinearFee.new(S.BigNum.from_str('44'), S.BigNum.from_str('155381')),
    // minimum utxo value
    S.BigNum.from_str('1000000'),
    // pool deposit
    S.BigNum.from_str('500000000'),
    // key deposit
    S.BigNum.from_str('2000000'),
    // max_value_size
    300,
    // max_tx_size
    8000
);
txBuilder.add_input(
    utxos[0].output().address(),
    utxos[0].input(),
    utxos[0].output().amount()
);
txBuilder.add_output(S.TransactionOutput.new(
      receiver,
      S.Value.new(
          S.BigNum.from_str('2000000')
      )
));
txBuilder.add_change_if_needed(
  changeAddr
);
  ```

**Comment by rooooooooob** [comment]
@CantTouchDis Thanks for the report. Just to verify 100%, the first profiling was on `8.1.0` and the second was `9.1.0`? And on the same browser/computer/input data?

>Ignoring the parameters max_value_size and max_tx_size as it happened in the previous version for me as well, what could be the issue?

I'm a bit confused since I thought we didn't merge in the change until `9.0.0` that made `max_value_size`/`max_tax_size` as arguments to `TransactionBuilder.new()`, or are you saying that you're just passing them in the first case even though they're not required in that version? Do your inputs that will be returned as change contain a lot of multiassets (e.g. NFTs)?

**Comment by CantTouchDis** [comment]
Sorry for describing it better.
Yes the first one was from `8.1.0` and the second one from the `9.1.0`.
In 8.1.0 I just had;
```
const txBuilder = S.TransactionBuilder.new(
    // all of these are taken from the mainnet genesis settings
    // linear fee parameters (a*size + b)
    S.LinearFee.new(S.BigNum.from_str('44'), S.BigNum.from_str('155381')),
    // minimum utxo value
    S.BigNum.from_str('1000000'),
    // pool deposit
    S.BigNum.from_str('500000000'),
    // key deposit
    S.BigNum.from_str('2000000')
    );
```

I wanted to say with that part of my post that I did not cause the problem, as it already happened in `8.1.0`

**Comment by rooooooooob** [comment]
Unfortunately I am not familiar with that profiling JSON data (i'm not a web dev either - I write rust code which we then do auto bindings for wasm/js from) and I tried to load it via the Performance tab -> import but it didn't do anything. That said, from looking at the pictures you attached and the code I think I can see what's taking so long.

I had asked about the version before since we reworked `TxBuilder::add_change_if_needed()` in `9.0.0` but only for cases with multiassets being returned as change. Your code wouldn't hit that case and all the CPU time seems to be coming just from the 3 calls (non-repeated) to `fees::min_fee()`. There are a couple crypto calls there which are taking all the time - namely to produce `fake_full_tx()` which we use for size-related calculations for the fee calculation.

Since I couldn't open the json I'm not sure if it's ` Bip32PrivateKey::from_bip39_entropy()` or `raw_key.sign([1u8; 100].as_ref())` or which crypto call since it's something calling the cryptoxide lib. @SebastienGllmt I'm guessing we could replace these with just raw fake bytes instead of computing it for our use case (size shouldn't change), right?

That said, these really shouldn't be taking that long - definitely not as long as you have encountered. Are other parts of the library also crazy slow for you? I would imagine that anything else involving crypto must be crazy slow too . Possibly this is just *really* bad performance for asm.js for the crypto libs. Is there any way you could try to use one of the non-asm.js bindings of the library and see if it's significantly faster there? I personally am not familiar with the asm.js bindings' performance.

tl;dr: I think asm.js performance might just be terrible for the crypto library we're using but this is insanely slow for what's happening.

**Comment by SebastienGllmt** [comment]
Yeah probably replacing `from_bip39_entropy` in the tx builder with some static result would speed it up

In general though asm.js is terribly slow for anything involving cryptography and shouldn't be used unless you have a platform limitation that requires it
